text/3151-scoped-threads.md

```diff
@@ -0,0 +1,237 @@
+handle2.join().unwrap();
+```
+
+作用域執行緒來解決！透過打開一個新的 `thread::scope()` 區塊，我們可以向編譯器證明在這作用域內產生的所有執行緒也會在這個作用域內結束生命。
+
+```rust
+let greeting = String::from("Hello world!");
+
+thread::scope(|s| {
+    let handle1 = s.spawn(|_| {
+        println!("thread #1 says: {}", greeting);
+    });
+
+    let handle2 = s.spawn(|_| {
+        println!("thread #2 says: {}", greeting);
+    });
+
+    handle1.join().unwrap();
+    handle2.join().unwrap();
+});
+```
+
+這意味著可以毫無顧忌地借用作用域之外的變數！
+
+現在我們不必再手動加入執行緒，因為所有未加入的執行緒將在作用域結束時自動會合：
+
+```rust
+let greeting = String::from("Hello world!");
+
+thread::scope(|s| {
+    s.spawn(|_| {
+        println!("thread #1 says: {}", greeting);
+    });
+
+    s.spawn(|_| {
+        println!("thread #2 says: {}", greeting);
+    });
+});
+```
+
+當以這種方式利用自動加入時，請注意，如果任何自動加入的執行緒出現恐慌，`thread::scope()` 將出現恐慌。
+
+你可能已經注意到作用域執行緒現在只接受一個參數，它只是對 `s` 的另一個引用。由於 `s` 存在於作用域內，我們不能直接借用它。使用傳遞的參數來生成巢狀執行緒：
+
+```rust
+thread::scope(|s| {
+    s.spawn(|s| {
+        s.spawn(|_| {
+            println!("I belong to the same `thread::scope()` as my parent thread")
+        });
+    });
+});
+```
+
+# 技術文件式解說
+[技術文件式解說]: #reference-level-explanation
+
+我們在 `std::thread` 模組中新增兩個新的型別：
+
+```rust
+struct Scope<'env> {}
+struct ScopedJoinHandle<'scope, T> {}
+```
+
+生命週期 `'env` 代表作用域外的環境，而 `'scope` 代表作用域本身。更準確地說，作用域外的所有內容都比 `'env` 和 `'scope` 內的所有內容都長。生命週期的關係是：
+
+```
+'variables_outside: 'env: 'scope: 'variables_inside
+```
+
+接下來，我們需要 `scoped()` 跟 `spawn()` 函式：
+
+```rust
+fn scope<'env, F, T>(f: F) -> T
+where
+    F: FnOnce(&Scope<'env>) -> T;
+
+impl<'env> Scope<'env> {
+    fn spawn<'scope, F, T>(&'scope self, f: F) -> ScopedJoinHandle<'scope, T>
+    where
+        F: FnOnce(&Scope<'env>) -> T + Send + 'env,
+        T: Send + 'env;
+}
+```
+
+這就是作用域執行緒的要點，真的。
+
+現在我們只需要再做兩件事來使 API 完整。首先，`ScopedJoinHandle` 等同於 `JoinHandle`，但與 `'scope` 生命週期掛勾，所以它將有同樣的方法。第二，執行緒生成器需要能夠在一個作用域内生成執行緒。
+
+```rust
+impl<'scope, T> ScopedJoinHandle<'scope, T> {
+    fn join(self) -> Result<T>;
+    fn thread(&self) -> &Thread;
+}
+
+impl Builder {
+    fn spawn_scoped<'scope, 'env, F, T>(
+        self,
+        &'scope Scope<'env>,
+        f: F,
+    ) -> io::Result<ScopedJoinHandle<'scope, T>>
+    where
+        F: FnOnce(&Scope<'env>) -> T + Send + 'env,
+        T: Send + 'env;
+}
+```
+
+# 缺點
+[缺點]: #drawbacks
+
+作用域執行緒的主要缺點是使標準函式庫有點大。
+
+# 基本原理和替代方案
+[基本原理和替代方案]: #rationale-and-alternatives
+
+* 將作用域執行緒保留在外部 crates 之中。
+
+  將他們放在標準函式庫有幾個優點：
+
+  * 這是一個非常常見和實用的工具，非常適合學習、測試和探索性程式設計。每個學習 Rust 的人都會在某個時候遇到借用和執行緒的互動。有一個非常重要的教訓是執行緒實際上**可以**借用局部變數，但標準函式庫並沒有反映這一點。
+
+  * 有些人可能會爭辯說我們應該完全不鼓勵使用執行緒，而是將人們指向像 Rayon 和 Tokio 這樣的執行器。但是，`thread::spawn()` 需要 F: `F: 'static` 並且無法繞過它，這感覺就像是標準函式庫中缺少的部分。
+
+  * 實現作用域執行緒非常難處理，因此最好有標準函式庫提供一個可靠的解決方案。
+
+  * 官方文檔和書籍中有許多範例可以透過作用域執行緒進行簡化。
+
+  * 作用域執行緒通常是比 `thread::spawn()` 更好的預設值，因為它們確保生成的執行緒被連接並且不會意外 「洩漏」。這有時是單元測試中的一個問題，如果單元測試產生執行緒並忘記加入它們，「迷途」 的執行緒可能會積累。
+
+  * 使用者一直在 IRC 和論壇上詢問作用域執行緒。將它們作為 `std::thread` 中的“祝福”模式對每個人都有好處。
```

Comment:
I missed this quotation mark. Should change to 「」 as well. Thanks!

Reply:
It's okay, this is my negligence